[1단계 - 로또 구현] 머랭(김대현) 미션 제출합니다.

cookie-meringue opened the pull request:
머랭 미션 제출합니다.

안녕하세요!
페어와 함께 구현하려니 생각보다 구현이 오래 걸리네요 ㅎㅎ
겨우 겨우 돌아가는 쓰레기를 완성했습니다 ㅠㅠ
그래서 그런지, 코드 스타일이 뒤죽박죽입니다..
제가 보기에도 만족스러운 코드는 아니지만, 리뷰해주시면 빠르게 반영하겠습니다!

++ 지금 확인했는데.. 코드 패키지 구조가 java/src/main/src 로 되어 있네요
리뷰 주시면 java/src/main/lotto 로 변경하겠습니다!

추가로, 질문이 생겨 남깁니다!

## 질문1: MVC 에서 View 가 Model 의 데이터를 사용하는 방법
View 에서는 Model 을 통해 나온 결과를 출력해야 한다고 알고 있습니다.
Lotto(Model) 번호를 View 가 출력해야 한다고 생각하는데요.
View 는 동시에 Lotto(Model) 에 대해 알면 안 된다고 알고 있습니다.

그렇게 되면, View 가 파라메터로 Lotto(Model)을 받을 수 없겠죠?
```
// View 의 파라메터에 Lotto(Model) 이 있으면 안됨!
void printLotto(Lotto lotto) {
...
}
```

이를 해결하기 위해 DTO 클래스인 ...Response 를 사용했는데요.
```
// Lotto 를 직접 다루지 않음.
void printLotto(LottoResponse lottoResponse) {
...
}
```

이런 설계에 대해 어떻게 생각하시는 지 궁금합니다! (Response 클래스들이 view 패키지 내부에 있는 점에 대한 생각도 궁금합니다!)


## 질문2: 패키지명 중복

상위에 lotto 패키지가 있다고 할 때, 하위 패키지 중 하나의 이름이 lotto 가 되어도 괜찮을까요?

예시:
java/src/main/lotto/model/lotto

우선 제 생각을 말씀드리자면
1. 패키지명이 중복된다고 해서 문제가 발생할 것 같지 않다.
2. model 내부에 Lotto 클래스를 추가한다고 할 때, lotto 라는 패키지명 말고는 마땅한 것이 없어 상위 패키지명과 같아도 된다.

입니다!




Prolog : https://prolog.techcourse.co.kr/studylogs/3869

cookie-meringue committed:
docs: 구현 기능 목록 작성

cookie-meringue committed:
feat: 애플리케이션 메인 클래스 작성

cookie-meringue committed:
chore: 불필요한 .gitkeep 제거

cookie-meringue committed:
feat: 입력 뷰 인터페이스 추가

cookie-meringue committed:
feat: 출력 뷰 인터페이스 추가

cookie-meringue committed:
feat: 콘솔 입력 뷰 추가

cookie-meringue committed:
feat: 콘솔 출력 뷰 추가

cookie-meringue committed:
feat: 컨트롤러 추가

cookie-meringue committed:
feat: 로또 구매 금액 입력 기능 구현

cookie-meringue committed:
feat: 로또 구매 금액 입력을 위한 출력 기능 구현

cookie-meringue committed:
test: ConsoleInputView 테스트 작성

- inputPurchaseMoney() 에 대한 테스트.

cookie-meringue committed:
feat: 로또 구매 금액을 입력받는 기능 구현

cookie-meringue committed:
docs: 로또 구매 금액 입력 기능 완료 표시

cookie-meringue committed:
docs: 구현 기능 목록 2, 3 통합

cookie-meringue committed:
docs: 구입 금액에 따라 로또 발행 기능 구현 목록 작성

cookie-meringue committed:
feat: 숫자 생성 인터페이스 추가

cookie-meringue committed:
feat: 기본 숫자 생성 클래스 구현

- NumberGenerator 의 구현체.

cookie-meringue committed:
test: DefaultNumberGenerator 테스트 작성

- generate() 에 대한 테스트.

cookie-meringue committed:
feat: 로또 번호 vo 구현

- 로또 번호 원시값을 포장.
- 로또 번호에 대한 validation 수행.
- 상수는 같은 패키지 사용을 위해 protected 로 설정.

cookie-meringue committed:
feat: LottoNumber 의 일급 컬렉션 구현

- NumberGenerator 를 통해 생성되는 정적 팩토리 생성자 구현.

cookie-meringue committed:
test: LottoNumber 테스트 작성

- 로또 번호 validation 을 테스트.

cookie-meringue committed:
feat: NumberGenerator 모킹 클래스 구현

- 설정한 숫자를 generate 하도록 모킹.

cookie-meringue committed:
feat: 로또 클래스 구현

- 로또 번호 일급 컬렉션인 LottoNumber 를 필드로 가짐.
- NumberGenerator 를 통해 생성될 수 있는 정적 팩토리 생성자 존재.

cookie-meringue committed:
feat: 로또 발행기 구현

- 구매 금액을 통해 구해진 수량만큼의 로또를 발행.
- NumberGenerator 를 DI 받아 사용.
- DI 는 Controller 에서 수행 예정.

cookie-meringue committed:
test: LottoMachine 테스트 작성

- 로또 구매 금액이 나누어떨어지지 않으면 예외 발생 여부 테스트.

cookie-meringue committed:
feat: 입력받은 로또 구입 금액을 통해 로또를 발행하는 기능 추가

- LottoMachine 을 이용.
- LottoMachine 에서 필요한 NumberGenerator 를 주입.

cookie-meringue committed:
feat: main 에서 LottoController 를 사용하도록 구현

cookie-meringue committed:
docs: 구입 금액에 따라 로또를 발행 기능 완료 표시

cookie-meringue committed:
docs: 발행받은 로또들을 출력 기능 구현 목록 작성

cookie-meringue committed:
feat: value 에 대한 getter 추가

cookie-meringue committed:
feat: values 에 대한 getter 추가

- 디미터 법칙에 의해 원시값을 반환해야 하므로, 필드들의 원시값을 리스트로 반환.

cookie-meringue committed:
feat: numbers 에 대한 getter 추가

- numbers 는 원시값을 반환하므로, 그대로 반환.

cookie-meringue committed:
feat: 로또 Response 클래스 작성

- Lotto 를 받아 자신의 numbers 를 초기화.
- Model 과 View 사이 의존성을 최소화하기 위한 작업.
- OutputView 에서 사용 예정.

cookie-meringue committed:
feat: 구입 로또 출력 함수 시그니처 추가

cookie-meringue committed:
feat: 구입 로또 출력 함수 구현

cookie-meringue committed:
feat: 발행받은 로또들을 출력 기능 구현

cookie-meringue committed:
docs: 발행받은 로또들을 출력 기능 완료 표시

cookie-meringue committed:
docs: 지난 주 당첨 번호를 입력 받는 기능 구현 목록 작성

cookie-meringue committed:
feat: 지난 주 당첨 번호를 입력 받기 위한 메시지 출력 함수 시그니처 추가

cookie-meringue committed:
feat: 지난 주 당첨 번호를 입력 받기 위한 메시지 출력 함수 구현

cookie-meringue committed:
feat: 지난 주 당첨 번호를 입력 받는 함수 시그니처 추가

cookie-meringue committed:
feat: 지난 주 당첨 번호를 입력 받는 함수 구현

cookie-meringue committed:
test: 지난 주 당첨 번호를 입력 받는 함수 테스트 작성

cookie-meringue committed:
feat: 지난 주 당첨 번호를 입력 받는 기능 구현

cookie-meringue committed:
docs: 지난 주 당첨 번호를 입력 받는 기능 완료 표시

cookie-meringue committed:
feat: 보너스 번호 입력을 위한 출력 함수 시그니처 추가

cookie-meringue committed:
feat: 보너스 번호 입력을 위한 출력 함수 구현

cookie-meringue committed:
feat: 보너스 번호 입력 함수 시그니처 추가

cookie-meringue committed:
feat: 보너스 번호 입력 함수 구현

cookie-meringue committed:
test: 보너스 번호 입력 함수 테스트 작성

cookie-meringue committed:
refactor: 복잡한 로직을 함수로 분리

cookie-meringue committed:
feat: 보너스 번호을 입력 받는 기능 구현

cookie-meringue committed:
docs: 보너스 번호을 입력 받는 기능 완료 표시

cookie-meringue committed:
docs: 로또 결과를 계산 기능 구현 목록 작성

cookie-meringue committed:
chore: 로또 관련 클래스들을 lotto 패키지로 분리

- 당첨 로또 생성으로 패키지 분리 필요.

cookie-meringue committed:
feat: equals, hashCode 재정의

- 컬렉션을 통한 비교를 위함.

cookie-meringue committed:
feat: LottoNumber 포함 여부 함수 추가

- 디미터 법칙을 위해 원시값을 받아 LottoNumber 로 포장하여 구현.
- 메모리적으로 이슈가 존재하므로, 추후 캐싱 도입 예정.

cookie-meringue committed:
feat: 보너스 번호 vo 구현

cookie-meringue committed:
feat: 당첨 로또 번호 vo 구현

cookie-meringue committed:
feat: 당첨 로또 번호 일급 컬렉션 구현

cookie-meringue committed:
test: 당첨 로또 번호 vo 테스트 작성

cookie-meringue committed:
test: 당첨 로또 번호 일급 컬렉션 테스트 작성

cookie-meringue committed:
test: 보너스 번호 vo 테스트 작성

cookie-meringue committed:
feat: 당첨 로또 클래스 구현

- 당첨 번호 일급 컬렉션, 보너스 번호 vo 를 가짐.

cookie-meringue committed:
docs: 당첨 로또 계산 시 발생 예외 요구사항 추가

cookie-meringue committed:
docs: 당첨 로또 계산 시 Overflow 예외처리를 TODO 로 변경

cookie-meringue committed:
test: 당첨 로또 클래스 테스트 작성

cookie-meringue committed:
feat: 로또 당첨 결과 Enum 구현

- Lotto 와 WinningLotto 를 사용하는 정적 팩토리 생성자 존재.

cookie-meringue committed:
test: 로또 당첨 결과 Enum 테스트 작성

cookie-meringue committed:
feat: 로또 당첨 결과 계산 메서드 구현

cookie-meringue committed:
feat: 로또 당첨 결과 계산 기능 구현

cookie-meringue committed:
docs: 로또 당첨 결과 계산 기능 완료 표시

cookie-meringue committed:
docs: 계산한 로또 결과 출력 기능 구현 목록 작성

cookie-meringue committed:
feat: 데이터 추출을 위한 getter 추가

cookie-meringue committed:
docs: 로또 결과 출력 출력 형식 올바르게 수정

- 공백 제거

cookie-meringue committed:
feat: 로또 당첨 결과 응답 Response 클래스 작성

- Lotto 를 받아 자신의 numbers 를 초기화.
- Model 과 View 사이 의존성을 최소화하기 위한 작업.
- OutputView 에서 사용 예정.

cookie-meringue committed:
feat: 계산한 로또 결과 출력 함수 시그니처 추가

cookie-meringue committed:
feat: 계산한 로또 결과 출력 함수 구현

- 적게 맞은 것 부터 출력해야 하기 때문에, Comparator 와 TreeMap 사용.
- 추후 함수, 로직 리팩터링 필요.

cookie-meringue committed:
feat: 계산한 로또 결과 출력 기능 구현

cookie-meringue committed:
feat: 계산한 로또 결과 출력 기능 완료 표시

cookie-meringue committed:
refactor: 사용하지 않는 getter 제거

cookie-meringue committed:
refactor: 의미 없는 중첩 Comparator 제거

practice2025 commented on README.md:
```diff
@@ -1,3 +1,66 @@
+
+### 4. 지난 주 당첨 번호를 입력받는다. (완료)
+
+- 입력하는 양식이 다르면 IllegalArgumentException 을 발생시킨 후, 애플리케이션을 종료한다.
```
👍

practice2025 commented on src/test/java/src/model/LottoPrizeTest.java:
```diff
@@ -0,0 +1,80 @@
+        );
+    }
+
+    @ParameterizedTest
```
ParameterizedTest 👍

practice2025 commented on src/main/java/src/model/LottoMachine.java:
```diff
@@ -0,0 +1,40 @@
+                .toList();
+    }
+
+    public List<LottoPrize> getLottoResults(List<Lotto> lottos, WinningLotto winningLotto) {
```
get보다 좀 더 의미를 드러내는 이름이 있지 않을까요?

cookie-meringue replied:
get 은 getter 와 혼동될 수 있고, 로또 결과를 계산해 응답한다는 의미에 부합하지 않는 네이밍인 것 같습니다 ㅎㅎ
감사합니다

practice2025 commented on src/main/java/src/model/winning_lotto/WinningLotto.java:
```diff
@@ -0,0 +1,55 @@
+        this.bonusNumber = bonusNumber;
+    }
+
+    public static WinningLotto of(List<Integer> numbers, int bonusNumber) {
```
정적 팩토리 메서드를 꼭 써야하는지 고민은 해보면 좋을 것 같습니다! 
변경은 하지 않아도 괜찮습니다

cookie-meringue replied:
정적 팩토리는 `Lotto.generateFrom()` 처럼 생성자에 의미를 부여할 수 있다는 이유로 사용한다고 알고 있습니다.
생성 방식이 하나인 경우 사용하기에 좋은 선택은 아니었던 것 같네요!
WinningLotto 의 of() 는 기본 생성자와 동일한 파라메터를 가지고 있기 때문에 좋은 방식은 아니었던 것 같습니다!

practice2025 commented on src/main/java/src/model/winning_lotto/vo/BonusNumber.java:
```diff
@@ -0,0 +1,35 @@
+
+public class BonusNumber {
+
+    protected static final int MIN_NUMBER_RANGE = 1;
```
RANGE가 아니라 MIN_NUMBER가 명확할 것 같습니다.

cookie-meringue replied:
감사합니다!

practice2025 commented on src/main/java/src/view/input/ConsoleInputView.java:
```diff
@@ -0,0 +1,63 @@
+
+            return parsedWinningLottoNumbers.stream().map(Integer::parseInt).toList();
+        } catch (RuntimeException e) {
+            throw new IllegalArgumentException("잘못된 입력입니다.", e);
```
익셉션을 넘겨주는 것 👍

practice2025 reviewed (changes requested):
안녕하세요 머랭! 첫 번째 미션 너무 고생많으셨습니다.
깔끔하게 잘 구현해주셨네요! 코멘트 몇개만 남겼습니다~

**질문1: MVC 에서 View 가 Model 의 데이터를 사용하는 방법**
정답은 없습니다. View에서 Lotto(Model)을 아는 것은 문제가 없습니다 (변경을 하지 않는다면).  Dto를 만드는 것은 결국 관계를 느슨하게 하는 것인데요 이것이 시스템이 커지고 복잡도가 높아지면 필요한 경우가 생기겠지만 지금의 요구사항에서 꼭 있어야 할지는 고민이 필요합니다. 얻는 이득이 없는데 코드양만 늘어난다면 굳이 만들 필요가 없겠죠? 

**질문2: 패키지명 중복**
일관성만 있다면 문제 없습니다~

cookie-meringue commented:
@practice2025 
감사합니다!
제 [질문1](https://github.com/woowacourse/java-lotto/pull/562#issue-2850327893)의 [답변](https://github.com/woowacourse/java-lotto/pull/562#pullrequestreview-2615612886)에 대해 추가 질문이 있습니다!

MVC 구조를 공부하는 과정에서 View 가 "모델이나 컨트롤러와 같이 다른 구성요소들을 몰라야 된다." 라고 표현하는 [글](https://m.blog.naver.com/jhc9639/220967034588)을 많이 접할 수 있었는데요.

View 가 Model 을 몰라야 한다면, 파라메터로도 Lotto(Model) 을 받으면 안 되는 것이 아닌지 의문이 들었습니다!

+ 그런데 [다른 글](https://www.youtube.com/watch?v=ogaXW6KPc8I) (5:16) 에서는 View 가 Model 만을 알아야 한다고 하더라구요..?
무엇이 맞는 말인지 헷갈립니다 ㅠㅠ

cookie-meringue committed:
refactor: 로또 결과 계산 함수명을 더 명확하게 변경

- get~ 네이밍은 getter 와 혼동될 가능성 존재.
- get 보다 더 명확한 의미의 함수명으로 변경.

cookie-meringue committed:
refactor: 로또 번호 상수 네이밍을 명확하게 변경

cookie-meringue committed:
refactor: 예외 발생 시 상수를 사용하도록 변경

- 예외 메시지에서 상수를 활용하도록 변경.

cookie-meringue committed:
refactor: 랜덤 번호 생성 코드 IntStream 적용

- 가독성을 위함.

cookie-meringue committed:
refactor: 사용되지 않는 getter 제거

cookie-meringue committed:
style: 코드 가독성 향상

cookie-meringue committed:
refactor: 정수 숫자 매칭 regex 상수로 분리

cookie-meringue committed:
refactor: 보너스 볼 일치에 따른 출력문 분기 로직 최적화

- 뷸필요햔 if 문 내부 코드 제거.

cookie-meringue committed:
rename: 프로젝트 패키지명 변경

- 이전의 실수로 인해 패키지명이 src 로 설정되어 있었음.
- lotto 로 변경함으로써 로또 프로젝트라는 것을 명시.

cookie-meringue committed:
refactor: 사용하지 않은 getter 제거

cookie-meringue committed:
command: Controller 에서 필요로 하는 의존성을 외부에서 주입하도록 변경

- Controller 가 의존성을 관리하는 것은 책임에 벗어났다고 판단.

cookie-meringue committed:
feat: 돈을 저장하는 vo 구현

cookie-meringue committed:
test: 돈을 저장하는 vo 테스트 작성

cookie-meringue committed:
refactor: 로또 구입 시 원시값 대신 Money 를 사용하도록 책임 분리

- Service 의 책임을 덜기 위함.
- 로또 금액을 검증하고 나누는 것은 LottoMachine 보단 vo 가 하는 것이 더 나은 책임 분리라고 판단.

practice2025 commented:
MVC는 규칙이라기보다는 아키텍처 패턴입니다. 
즉, “View가 Model을 몰라야 한다”는 원칙은 권장사항(Best Practice)이지 무조껀 따라야 하는 법칙은 아닙니다. 
만약 머랭이 혼자 개발을 한다면 Model에 어떤 변경을 하면 이것이 View에 어떤 영향을 미치는지 다 알고 있을 것이라 생각합니다. 
그렇다면 어쩌면 굳이 코드레벨에서 분리를 안해도 될 수 있습니다. 하지만 작업자가 많아지고 시스템이 커질수록 별도의 뷰 모델을 만든다거나 Dto로 MVC사이의 의존성을 줄인다거나 하는 선택을 할 수 있겠지요. 만약 머랭이 미래에 시스템이 커질 것을 가정하고 분리를 했다고 해도 납득 했을 것이고 일단 당장 필요하지 않으니 직접 의존을 하게 했다 해도 납득 했을 것 입니다.

중요한 것은 의도로 어떤 선택을 했고 그것의 트레이드 오프에 대해 잘 알고 설명하는 것이라 생각합니다. 제가 말씀드린 것도 충분한 이해를 하고 의도를 가지고 만들었는지에 대한 이야기였습니다. 아마 앞으로의 과정에서도 그런 머랭의 선택과 의도에 대해 이야기를 주시면 더 많은 것들을 나눌 수 있을 것 같습니다.

cookie-meringue commented:
기술이나 아키텍처를 설계할 때 상황에 따라 설계하는 것이 중요하다는 것을 알게 되었네요 ㅎㅎ
어떻게 보면 틀에 갇혀있었는데 꺼내주셔서 감사합니다!

practice2025 reviewed (approved):
고생 많으셨어요 머랭!
다음 단계도 화이팅입니다~
